### PR TITLE
Add shrink expand dbm

### PR DESCRIPTION
**Pull request by frederikja163**
Added dbm_shrinkExpand from the UDBM library along with 2 auxilary functions.
`compute_tables` https://github.com/UPPAALModelChecker/UDBM/blob/main/src/dbm.c#L1713
`update_dbm` https://github.com/UPPAALModelChecker/UDBM/blob/main/src/dbm.c#L1767
`shrink_expand` https://github.com/UPPAALModelChecker/UDBM/blob/main/src/dbm.c#L1830

### CODE REVIEW

**Review by SilverBrother — COMMENTED** [review]
assert_eq! Lines 683 and 745, I assume is there to cause unhandled panics if the tables aren't the same size.
Lines 689 - 694 "compute_tables" does not align with diagrams presented. False pushes 0 rather than j and True pushes src_ind rather than 0 in the code.
Otherwise LGTM
update_dbm is the reason for the actual values in the expanded table you tested on the whiteboard yesterday
Sorry for unprofessional comment <3

**Comment by frederikja163** [comment]
> assert_eq! Lines 683 and 745, I assume is there to cause unhandled panics if the table

File: src/zones/dbm.rs [hunk 1]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
         unsafe { dbm.assert_valid() }
     }
 
+    fn compute_tables(&self, src_clocks: &Vec<bool>, dst_clocks: &Vec<bool>) -> (Vec<ClockIndex>, Vec<ClockIndex>) {
+        let mut dst_to_src = Vec::<ClockIndex>::default();
+        let mut src_to_dst = Vec::<ClockIndex>::default();
+        let mut src_ind = 0;
+        assert_eq!(src_clocks.len(), dst_clocks.len());
```
Comment (SilverBrother) [inline src/zones/dbm.rs:684]: unhandled panic. intentional?

File: src/zones/dbm.rs [hunk 2]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+        for (src, dst) in src_clocks.iter().zip(dst_clocks.iter()){
+            let (src, dst) = (src.to_owned(), dst.to_owned());
+            if dst {
+                src_to_dst.push(dst_to_src.len());
+                if src {
+                    dst_to_src.push(src_ind);
+                }
+                else {
+                    dst_to_src.push(0);
+                }
```
Comment (SilverBrother) [inline src/zones/dbm.rs]: Diagram shows the opposite

File: src/zones/dbm.rs [hunk 3]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+                for j in 0..self.dim {
+                    self.data[j + i * self.dim] = LS_INFINITY;
+                }
+            }
+            self.data[i + i * self.dim] = LE_ZERO;
+        }
+    }
+
+    fn shrink_expand(&self, src_clocks: &Vec<bool>, dst_clocks: &Vec<bool>) -> (DBM<Valid>, Vec<ClockIndex>) {
+        assert_eq!(src_clocks.len(), dst_clocks.len());
```
Comment (SilverBrother) [inline src/zones/dbm.rs:765]: unhandled panics. intentional?

File: src/zones/dbm.rs [hunk 4]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+
+        for i in 1..self.dim {
+            if dst_to_src[i] != 0 { // If copy from src.
+                let constraint0 = src.data[src.dim * dst_to_src[i]];
+                self.data[i * self.dim] = constraint0;
+
+                for j in 1..self.dim {
+                    if dst_to_src[j] == 0
+                    {
+                        self.data[j + i * self.dim] = constraint0;
```
Comment (SilverBrother) [inline src/zones/dbm.rs:731]: Thinks this is the cause for the numbers that arent infinity when you expand

File: src/zones/dbm.rs (hunk 2 above)
Comment (frederikja163) [reply src/zones/dbm.rs]: I think that is a fault of the diagram in that case. I did have some problems with if statements and their boolean expressions. I think our report is gonna have a more thorough proof of the conversion from UDBM than what i have made to this point.

File: src/zones/dbm.rs (hunk 4 above)
Comment (frederikja163) [reply src/zones/dbm.rs:731]: Yes, i am aware of that. I dont know why UDBM decided to do it like this though. Thats what i was talking about at the meeting yesterday.

File: src/zones/dbm.rs (hunk 3 above)
Comment (frederikja163) [reply src/zones/dbm.rs:765]: Yes, if these two arent equal the program wouldn't behave as expected.

File: src/zones/dbm.rs (hunk 1 above)
Comment (frederikja163) [reply src/zones/dbm.rs:684]: Yes, if these two arent equal the program wouldn't behave as expected.

File: src/zones/dbm.rs (hunk 2 above)
Comment (SilverBrother) [reply src/zones/dbm.rs]: If the diagram is off, then the code LGTM

File: src/zones/dbm.rs (hunk 4 above)
Comment (SilverBrother) [reply src/zones/dbm.rs:731]: Not to make it an entire topic, but in the process of expanding, when the new row is created the for loop still iterates over the other axis, i suspect that it carries this line over to the expanded dimension and overwrites the infinity. Just a thought, but you have my stamp of approval

**Review by SilverBrother — APPROVED** [review]

File: src/zones/dbm.rs (hunk 4 above)
Comment (frederikja163) [reply src/zones/dbm.rs:731]: No, this line only gets run if the clock is in the src. There should be an if statement pretty much directly above the for loop in your comment.

File: src/zones/dbm.rs [hunk 5]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
     }
 
+    fn compute_tables(&self, src_clocks: &Vec<bool>, dst_clocks: &Vec<bool>) -> (Vec<ClockIndex>, Vec<ClockIndex>) {
+        let mut dst_to_src = Vec::<ClockIndex>::default();
+        let mut src_to_dst = Vec::<ClockIndex>::default();
+        let mut src_ind = 0;
+        assert_eq!(src_clocks.len(), dst_clocks.len());
+
+        for (src, dst) in src_clocks.iter().zip(dst_clocks.iter()){
+            let (src, dst) = (src.to_owned(), dst.to_owned());
```
Comment (Aavild) [inline src/zones/dbm.rs]: pretty sure this is unneccesary

File: src/zones/dbm.rs [hunk 6]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+            if dst {
+                src_to_dst.push(dst_to_src.len());
+                if src {
```
Comment (Aavild) [inline src/zones/dbm.rs]: I would find this more readable if the inner if statement is a conditional expression instead, but this is fine

Comment (Aavild) [reply src/zones/dbm.rs]: that is `dst_to_src.push(if src { src_ind} else { 0 })`

File: src/zones/dbm.rs [hunk 7]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
         unsafe { dbm.assert_valid() }
     }
 
+    fn compute_tables(&self, src_clocks: &Vec<bool>, dst_clocks: &Vec<bool>) -> (Vec<ClockIndex>, Vec<ClockIndex>) {
```
Comment (Aavild) [inline src/zones/dbm.rs:680]: example:
[T, T, F, T, T, F, F, T] - dst
[T, F, T, T, T, F, F, T] - src
[0, 1, 0, 2, 3, 0, 0, 4] - src_to_dst
[0, 0, ., 3, 4, ., ., 7] = [0, 0, 3, 4, 7] - dst_to_src

Comment (Aavild) [reply src/zones/dbm.rs:680]: shouldn't it be the following?:
[T, T, F, T, T, F, F, T] - dst
[T, F, T, T, T, F, F, T] - src
[0, 0, 1, 2, 3, 0, 0, 4] - src_to_dst
[0, ., 2, 3, 4, ., ., 7] = [0, 2, 3, 4, 7] - dst_to_src

File: src/zones/dbm.rs [hunk 8]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+            src_ind += 1;
+        };
+
+        return (src_to_dst, dst_to_src);
+    }
+
+    fn update_dbm(&mut self, src: &DBM<Valid>, dst_to_src: &Vec<ClockIndex>) {
+        self.data[0] = LE_ZERO;
+
+        for j in 1..self.dim {
```
Comment (Aavild) [inline src/zones/dbm.rs:711]: this loop should explain that these are the constraints for clock [0,j]

File: src/zones/dbm.rs [hunk 9]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+            else
+            {
+                self.data[j] = src.data[dst_to_src[j]];
+            }
+        }
+
+        for i in 1..self.dim {
+            if dst_to_src[i] != 0 { // If copy from src.
+                let constraint0 = src.data[src.dim * dst_to_src[i]];
+                self.data[i * self.dim] = constraint0;
```
Comment (Aavild) [inline src/zones/dbm.rs:726]: simillar to clock [0,j] this should explain it is [i, 0]

File: src/zones/dbm.rs (hunk 3 above)
Comment (Aavild) [inline src/zones/dbm.rs:765]: maybe a debug assert for counting that src doesn't have more true values than dst?

File: src/zones/dbm.rs [hunk 10]
```diff
@@ -676,6 +676,79 @@ impl DBM<Valid> {
...
+                    self.data[j + i * self.dim] = LS_INFINITY;
+                }
+            }
+            self.data[i + i * self.dim] = LE_ZERO;
+        }
+    }
+
+    fn shrink_expand(&self, src_clocks: &Vec<bool>, dst_clocks: &Vec<bool>) -> (DBM<Valid>, Vec<ClockIndex>) {
+        assert_eq!(src_clocks.len(), dst_clocks.len());
+        let (src_to_dst, dst_to_src) = self.compute_tables(src_clocks, dst_clocks);
```
Comment (Aavild) [inline src/zones/dbm.rs:766]: I don't think it is that efficient to calculate remove_clock([2,3]) to src_clocks([T, T, F, F, T]) to then update_dbm([0, 1, 4])

File: src/zones/dbm.rs [hunk 11]
```diff
@@ -1296,6 +1369,9 @@ impl<T: DBMState> Display for DBM<T> {
 
 #[cfg(test)]
```
Comment (Aavild) [inline src/zones/dbm.rs:1390]: make a comment showing how they work. example for dbm_shrink_expand
// 0, 1, 2, 3, 4
// 5, 6, 7, 8, 9
// 10, 11, 12, 13, 14
// 15, 16, 17, 18, 19
//, 20, 21, 22, 23, 24
is redrawn into
// 0, 1, 4, inf, inf
// 5, 0, 9, inf, inf
// 10, 11, 0, inf, inf
// 0, 0, 0, 0, inf
//  0, 0, 0, 0, 0

**Review by Aavild — CHANGES_REQUESTED** [review]
maybe update rust to newest version with `rustup update` and then `cargo clippy --all-features --all-targets` and then `cargo fmt --all`?

File: src/zones/dbm.rs (hunk 5 above)
Comment (frederikja163) [reply src/zones/dbm.rs]: It isn't, the for loop will give us a borrowed value, which we need to convert to an owned value.

File: src/zones/dbm.rs (hunk 6 above)
Comment (frederikja163) [reply src/zones/dbm.rs]: Yeah, i had it like that at first, but i found this to be preferable as its easier to debug.

File: src/zones/dbm.rs (hunk 10 above)
Comment (frederikja163) [reply src/zones/dbm.rs:766]: Im not sure what you mean? By your comment

Comment (frederikja163) [reply src/zones/dbm.rs:766]: Besides, this is the algorithm implemented by UPAAL i think we should strive to copy them as closely as possible

File: src/zones/dbm.rs (hunk 7 above)
Comment (frederikja163) [reply src/zones/dbm.rs:680]: I cant really make sense of what is happening in the bottom example. But the first one is correct and what i would expect it to be.

File: src/zones/dbm.rs (hunk 3 above)
Comment (frederikja163) [reply src/zones/dbm.rs:765]: We could put many debug asserts in here to check if all the parameters are correct. But ultimately i think its better if the user has the responsibility to make sure the parameters are correct.

**Review by Aavild — APPROVED** [review]

**Pull request merged**